pkg/component/runtime/service_command_test.go

```diff
@@ -181,3 +195,162 @@ func TestExecuteCommand(t *testing.T) {
+			cmdCtx, log, exePath, &component.ServiceOperationsCommandSpec{},
+			retryCtx, defaultRetrySleepInitDuration, retrySleepMaxDuration,
+		)
+		require.NoError(t, err)
+
+		<-retryCtx.Done()
+		checkRetryLogs(t, observer, exeConfig)
+	})
+
+	// Execution failed - with retry configuration in spec
+	t.Run("failed_execution_with_retry_config", func(t *testing.T) {
+		cmdCtx := context.Background()
+		log, observer := logger.NewTesting(t.Name())
+
+		exeConfig := progConfig{
+			ErrMessage: "foo bar",
+			ExitCode:   111,
+		}
+		exePath, err := prepareTestProg(cmdCtx, log, t.TempDir(), exeConfig)
+		require.NoError(t, err)
+
+		// Since the service command is retried indefinitely, we need a way to
+		// stop the test within a reasonable amount of time
+		retryCtx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+		defer cancel()
+
+		defaultRetrySleepInitDuration := 50 * time.Millisecond
+		retrySleepMaxDuration := 1 * time.Second
+
+		spec := &component.ServiceOperationsCommandSpec{
+			// We deliberately set RetrySleepInitDuration to just shorter
+			// than the retryCtx timeout. With this we should observe:
+			// - the initial execution of the command (before any retries)
+			// - one message about the next (first) retry
+			// - one more execution of the command, as a result of the first retry
+			// - one message about the next (second) retry
+			RetrySleepInitDuration: 700 * time.Millisecond,
+		}
+		err = executeServiceCommandWithRetries(
+			cmdCtx, log, exePath, spec,
+			retryCtx, defaultRetrySleepInitDuration, retrySleepMaxDuration,
+		)
+		require.NoError(t, err)
+
+		<-retryCtx.Done()
+		checkRetryLogs(t, observer, exeConfig)
+	})
+
+	// Execution succeeded after a few retries
+	t.Run("succeed_after_retry", func(t *testing.T) {
+		cmdCtx := context.Background()
+		log, observer := logger.NewTesting(t.Name())
+
+		const succeedCmdAfter = 2 * time.Second
+		now := time.Now()
+		exeConfig := progConfig{
+			ErrMessage:   "foo bar",
+			ExitCode:     111,
+			SucceedAfter: now.Add(succeedCmdAfter).UnixMilli(),
+		}
+		exePath, err := prepareTestProg(cmdCtx, log, t.TempDir(), exeConfig)
+		require.NoError(t, err)
+
+		// Since the service command is retried indefinitely, we need a way to
+		// stop the test within a reasonable amount of time. However, we should never
+		// hit this timeout as the command should succeed before the timeout is reached.
+		retryCtx, cancel := context.WithTimeout(context.Background(), 4*time.Second)
+		defer cancel()
+
+		defaultRetrySleepInitDuration := 50 * time.Millisecond
+		retrySleepMaxDuration := 1 * time.Second
+
+		spec := &component.ServiceOperationsCommandSpec{
+			RetrySleepInitDuration: 200 * time.Millisecond,
+		}
+		err = executeServiceCommandWithRetries(
+			cmdCtx, log, exePath, spec,
+			retryCtx, defaultRetrySleepInitDuration, retrySleepMaxDuration,
+		)
+		require.NoError(t, err)
+
+		// Give the command time to succeed.
+		time.Sleep(succeedCmdAfter)
```

Comment:
Not a fan of this `time.Sleep` here but I haven't come up with a better solution without adding more complexity to the code path being tested.  Open to suggestions!

Reply:
Can it be replaced with a call to https://pkg.go.dev/github.com/stretchr/testify/require#Eventually?

Reply:
Yeah I agree an eventually check might be safer.

Reply:
I did try to use `require.Eventually` but couldn't come up with a good boolean condition for the check.  That's why I ended up resorting to a `time.Sleep` instead 😞.  Any ideas?

Reply:
You could write a variant of `checkRetryLogs` that polls for a log message indicating the service command executed successfully.